sigstore/_internal/rekor/_client.py

```diff
@@ -2,8 +2,90 @@
 Client implementation for interacting with Rekor.
 """
 
+import json
+from abc import ABC
+from typing import Any, Optional, Tuple
+from urllib.parse import urljoin
+
+import requests  # type: ignore
+
+DEFAULT_REKOR_URL = "https://rekor.sigstore.dev/api/v1/"
+
+
+class Endpoint(ABC):
+    def __init__(self, url: str, session: requests.Session) -> None:
+        self.url = url
+        self.session = session
+
 
 class RekorClient:
     """The internal Rekor client"""
 
+    def __init__(self, url: str = DEFAULT_REKOR_URL) -> None:
+        self.url = url
+        self.session = requests.Session()
+        self.session.headers.update(
+            {"Content-Type": "application/json", "Accept": "application/json"}
+        )
+
+    @property
+    def index(self) -> Endpoint:
+        return RekorIndex(urljoin(self.url, "index/"), session=self.session)
+
+    @property
+    def log(self) -> Endpoint:
+        return RekorLog(urljoin(self.url, "log/"), session=self.session)
+
+
+class RekorIndex(Endpoint):
+    @property
+    def retrieve(self) -> Endpoint:
+        return RekorRetrieve(urljoin(self.url, "retrieve/"), session=self.session)
+
+
+class RekorRetrieve(Endpoint):
+    def post(self, sha256_hash: Optional[str] = None) -> requests.Response:
```

Comment:
We should figure out what the return type of these should be.

It should either work like this where we return a `requests.Response` which the caller is responsible for checking the status and then extracting the contents. Or we should do the checking within these functions and return something more specific to the endpoint (like the one below that returns `Tuple[str, str]`).

Reply:
I like what you're doing in https://github.com/trailofbits/sigstore-python/pull/13/ where these return dataclasses, we can do that here as well.

Reply:
Sure, I'll do this before merging.